### ISSUE TITLE
undefined method `embed` (NoMethodError) when calling screenshot_and_raise

### ISSUE THREAD

**Issue by juandelgado** [calabash/calabash-ios#246]
Hello there,

We are having some trouble with Calabash taking screenshots when something fails. Errors look like this:

```undefined method `embed' for #<MyAccount:0x007fa7f60c62c8> (NoMethodError)```

MyAccount being one of our classes.

Digging a bit in the code isn't [this line](https://github.com/calabash/calabash-ios/blob/0.9.x/calabash-cucumber/lib/calabash-cucumber/tests_helpers.rb#L107) assuming we are extending / using Calabash::IBase? Because we are not. And the problem is that because calling ```embed``` fails, message is never raised, making the debugging harder.

Also, shouldn't we have a simple way to disable screenshots altogether? I want to remember that this was possible (maybe on Calabash Android, environment var?), but looking at the code I can't see an easy way.

Thanks.

**Comment by jmoody** [comment]
Reproduced!

Ack! This one is nasty.  Apologies to everyone.  I will try to get this sorted quickly.

**Comment by TeresaP** [comment]
Awesome, thanks jmoody!

**Comment by jmoody** [comment]
**TL;DR**  There is a problem.  I have reproduced it in calabash-ios and calabash-android.  The problem is not easy to solve.  I will have to iterate with the other members of the Calabash team.  If you can see a nice fix, I encourage you to submit a pull request or comment here.

The problem boils down to this:  we are assuming that the `embed` method will be forwarded to `Cucumber::RbSupport::RbWorld#embed`.  The cucumber runtime will then figure out if any formatters are attached that can handle embedding.  If no such formatters are attached, `embed` becomes a `nop`.

In the context of a cucumber Step definition, `screenshot_and_raise` will work because `embed` is in the Cucumber world.  Put another way, `screenshot_and_raise` is being called in the context of RbWorld which defines `embed`.

In the context of the POM, this works because of these two methods:

```
# IBase / ABase
def page(clz, *args)
  # At runtime, `world` evals to self and self contains the RbWorld.
  # See below for the output of page(SamplePage).
  clz.new(world, *args)
end

def method_missing(name, *args, &block)
  world.send(name, *args, &block)
end
```

In concrete terms, `method_missing` is called with `:embed`, which is then forwarded to the page instance variable `@world`, which contains `RbWorld` which defines `embed`.

If you are not using ABase or IBase, then your classes and Step definitions probably look something like this:

```
module MyApp
  class HomePage
    include Calabash::Android::Operations

    def some_method
      screenshot_and_raise 'Hey!'
    end
    
  end
end

Then(/^I call some method on my home page$/) do
   page = MyApp::HomePage.new
   page.some_method
end
```

This is completely natural. The problem is that `embed` is not defined anywhere in MyApp::HomePage.

I think this is a bug.  We need to find a way to fix this.

Here are several workarounds.

### Monkey patching

In your `support/env.rb`, define an empty `embed` method on Calabash.  Obviously, you won't get screenshots embedded.  I am not sure what effect this will have on the XTC.

```
# Android
module Calabash
  module Android
    module Operations
      def embed(_, _, _)

      end
    end
  end
end

# iOS
module Calabash
  module Cucumber
    module FailureHelpers
      def embed(_,_,_)

      end
    end
  end
end
```

### Do what IBase/ABase does.

This pattern will break down quickly if you are using helper classes that eventually call `embed`.

```
module MyApp
  class BasePage

    attr_accessor :world

    def initialize(world)
      @world = world
    end

    def method_missing(name, *args, &block)
      world.send(name, *args, &block)
    end
    
  end
end

module MyApp
  class LoginPage < MyApp::BasePage

    attr_accessor :username
    attr_accessor :password
    def initialize(world, username, password)
      super(world)
      @username = username
      @password = password
    end

    def login
      screenshot_and_raise 'Hey!'
    end

  end
end

Then(/^I login at "([^"]*)" with "([^"]*)"$/) do |username, password|
  login_page = MyApp::LoginPage.new(self, username, password)
  login_page.login
end
```
    

#### page(SamplePage)

```
#<MyApp::SamplePage:0x007f7f6f410c30
 @transition_duration=0.5,
 @world=
  #<Object+Calabash::Android::Operations+
           Calabash::Android::WaitHelpers+
           Calabash::Android::TouchHelpers+
           Calabash::Android::Gestures+
           Calabash::Android::TextHelpers+
           Calabash::Android::EnvironmentHelpers+
           Calabash::Android::ColorHelper+
           Cucumber::RbSupport::RbWorld+ <=== embed is defined here.
           PP::ObjectMixin+
           JSON::Ext::Generator::GeneratorMethods::
           Object+Kernel:0x3fbfb61980ec>>
```

@TobiasRoikjer @krukow @jescriba

**Comment by jmoody** [comment]
The fix is working its way through CI.

It is important that your classes include `Calabash::Cucumber::Operations` and not  other modules within Calabash.

In fact, I am going to out on a limb here and say, if you are including _any other module than `Calabash::Cucumber::Operations` you are probably doing something wrong._  The Operations module _is_ the public API for Calabash; everything else should be considered private.

###### Bad

```
module NotPOM
  class HomePage
    include Calabash::Cucumber::FailureHelpers

    # Will raise undefined method `embed' error.
    def my_exceptional_method
      screenshot_and_raise 'Hey!'
    end

  end
end
```

###### Good

```
module NotPOM
  class HomePage
    include Calabash::Cucumber::Operations

    def my_exceptional_method
      screenshot_and_raise 'Hey!'
    end

  end
end
```

**Comment by TeresaP** [comment]
@adzellman this affects us but I can't remember if we need to change anything. Pleas let Andy know as I dont know if he checks his public Github account :)
